bevy_reflect: Adding support for Atomic values

recatek opened the pull request:
Fixes #14418

Note that this does not add AtomicPtr, which would need its own special casing support, just the regular value types.
Also, I was forced to be opinionated about which Ordering to use, so I chose SeqCst as the strictest by default.

recatek commented:
Looks like the build is failing due to something in bevy_render.

MrGVSV commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -220,6 +220,171 @@ impl_type_path!(::bevy_utils::NoOpHash);
+                $ty: ::core::any::Any + ::core::marker::Send + ::core::marker::Sync,
+            {
+                fn get_type_registration() -> bevy_reflect::TypeRegistration {
+                    let mut registration = bevy_reflect::TypeRegistration::of::<Self>();
+                    registration.insert::<bevy_reflect::ReflectFromPtr>(bevy_reflect::FromType::<
+                        Self,
+                    >::from_type());
+                    registration.insert::<bevy_reflect::ReflectFromReflect>(
+                        bevy_reflect::FromType::<Self>::from_type(),
+                    );
+                    registration
+                        .insert::<ReflectDefault>(bevy_reflect::FromType::<Self>::from_type());
+                    registration
```
Do the atomic types also support serde traits? If so, we could also register `ReflectSerialize` and `ReflectDeserialize`.

MrGVSV commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -220,6 +220,171 @@ impl_type_path!(::bevy_utils::NoOpHash);
+                }
+            }
+
+            impl_type_path!($ty);
```
Nit: I personally almost missed this call to `impl_type_path!`. Maybe we could move such macro invocations to the top/bottom of this scope?

MrGVSV replied:
We probably also want to include something like:
```rust
#[cfg(feature = "functions")]
crate::func::macros::impl_function_traits!($ty);
```

MrGVSV commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -220,6 +220,171 @@ impl_type_path!(::bevy_utils::NoOpHash);
+    ($ty:ty, $ordering:ident) => {
+        const _: () = {
+            #[allow(unused_mut)]
+            impl bevy_reflect::GetTypeRegistration for $ty
```
Since we only call this macro in this file, could we remove the fully-qualified type paths? Mainly just for consistency with the other macros and for readability.

MrGVSV reviewed (commented):
I think `SeqCst` is the right choice here, but I'm not familiar enough with atomics to fully make that decision. If someone else has feedback there, please feel free to leave a comment!

Otherwise just a couple suggestions!

vacuus commented:
Relaxed ordering should be fine. Those methods don't seem like points for synchronization, especially because they would need to synchronize with a store somewhere.

recatek commented:
I think I've addressed all of the feedback. Regarding Relaxed vs. SeqCst, SeqCst is the strictest option and is the default in other languages when you don't specify one. The only downside is pretty minor perf hits, and even then primarily on ARM rather than x86 (which has strong memory guarantees to begin with). Since this is in reflection code to begin with and because the user won't be able to control the ordering mode, I don't think the perf hit is as bad as potentially unexpected behavior.

recatek commented:
Is this still S-Needs-Review? I'm new to bevy PRs so not sure how all the labels work.

MrGVSV commented:
> Is this still S-Needs-Review? I'm new to bevy PRs so not sure how all the labels work.

Yeah you need 2 community approvals before getting moved to `S-Ready-For-Final-Review` :)
